lib/matplotlib/legend.py

```diff
@@ -1108,6 +1115,12 @@ def _find_best_position(self, width, height, renderer, consider=None):
         assert self.isaxes
 
         verts, bboxes, lines, offsets = self._auto_legend_data()
+        if self._loc_used_default and verts.shape[0] > 200000:
```

Comment:
The logic seems a bit convoluted. Do we really need a private variable `self._loc_used_default` and an additional `is_initial_setting` argument? AFAICS the warning is only issued with `is_initial_setting=True`, and that's only called in `__init__`. So the warning could be moved there.

Reply:
I'm not sure if we _need_ it, but here are the reasons I did it that way:

1. Someone could create the legend then change its `_loc` attribute before it is drawn. They would probably not like a warning being created.
2. Creating the warning in the `__init__` section would require calling `self._auto_legend_data()` twice. This function call may be expensive, but more importantly with all of the side effects I have a hard time guaranteeing that this function would return the same things if it's called in two different places. I assume there was a reason it is delayed until draw-time in the current implementation?

Reply:
Confirmed that if I put the warning in the `__init__` call, no warning is issued when the legend is created before some of the lines in a plot. In the below example, I modified the code to put the warning in the `__init__` call and the warning is only issued if `legend()` is called _after_ creating the large line plot. In the existing implementation, the warning is issued in either case. Both of the below blocks of code result in the same rendering.

```python
In [13]: f, ax = plt.subplots() 
    ...: ax.plot([0.5, 0, 1], label='small') 
    ...: ax.plot(np.random.random(200001)) 
    ...: ax.legend() 
    ...: plt.show()                                                                                                                                                
/home/kevin/miniconda3/envs/mpl-dev/bin/ipython:6: UserWarning: Creating legend with loc="best" can be slow with large amounts of data.

In [14]: f, ax = plt.subplots() 
    ...: ax.plot([0.5, 0, 1], label='small') 
    ...: ax.legend() 
    ...: ax.plot(np.random.random(200001)) 
    ...: plt.show()
```

This comes from the fact that `_auto_legend_data()` is iterating over the axes' lines.

Reply:
That being said, this is just a warning. It may be a reasonable tradeoff to simplify implementation in favor of sometimes not issuing a warning.

Reply:
Thanks for the clarification. Indeed, we shouldn't call `self._auto_legend_data()` twice just to check the warning condition. And from a conceptual point of view the warning is better placed in `_find_best_position()`.

See https://github.com/timhoffm/matplotlib/commit/917b5205645af78ea2a419293d147e97cce21422 for an alternative implementation with the same functionality. There's two changes:

1. `self._loc_used_default = loc is None` is better than splitting the definition in the `if else`.
2. One can get rid of the `is_initial_setting` by explicitly working around the change of `_set_loc` in the constructor. The workaround is not 100% nice as well because it has to know that `_set_loc` changes `_loc_used_default`. Nevertheless I find this more clear than defining an extra argument for that case.

Reply:
Thanks for the suggestion. I agree with your assessment that it's not 100% nice, but seems better than what I had before. Incorporated.